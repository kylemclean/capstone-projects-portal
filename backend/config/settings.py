"""
Django settings for config project.

Generated by 'django-admin startproject' using Django 3.2.7.

For more information on this file, see
https://docs.djangoproject.com/en/3.2/topics/settings/

For the full list of settings and their values, see
https://docs.djangoproject.com/en/3.2/ref/settings/
"""

import os
from pathlib import Path
from urllib.parse import urlparse

from environ import Env

# This is the backend directory
# Build paths inside the project like this: BASE_DIR / 'subdir'.
BASE_DIR = Path(__file__).resolve().parent.parent

# reading .env file
env = Env()
env.read_env(env_file=BASE_DIR / ".env")

# Quick-start development settings - unsuitable for production
# See https://docs.djangoproject.com/en/3.2/howto/deployment/checklist/

# Raises django's ImproperlyConfigured exception if SECRET_KEY not in os.environ
SECRET_KEY = env("DJANGO_SECRET_KEY")

# False if not in os.environ
DEBUG = env("DJANGO_DEBUG", default=False)

FRONTEND_BASE_URL = env("FRONTEND_BASE_URL", default="http://localhost:3000")
parsed_frontend_base_url = urlparse(FRONTEND_BASE_URL)

frontend_host = parsed_frontend_base_url.hostname
ALLOWED_HOSTS = [
    frontend_host,
] + (["localhost", "127.0.0.1", "[::1]"] if DEBUG else [])


# Application definition

INSTALLED_APPS = [
    "rest_framework",
    "rest_framework.authtoken",
    "django.contrib.admin",
    "django.contrib.auth",
    "django.contrib.contenttypes",
    "django.contrib.sessions",
    "django.contrib.messages",
    "django.contrib.staticfiles",
    "portal",
    "corsheaders",
    "drf_spectacular",
    "django_admin_listfilter_dropdown",
]

MIDDLEWARE = [
    "corsheaders.middleware.CorsMiddleware",
    "django.middleware.security.SecurityMiddleware",
    "django.contrib.sessions.middleware.SessionMiddleware",
    "django.middleware.common.CommonMiddleware",
    "django.middleware.csrf.CsrfViewMiddleware",
    "django.contrib.auth.middleware.AuthenticationMiddleware",
    "django.contrib.messages.middleware.MessageMiddleware",
    "django.middleware.clickjacking.XFrameOptionsMiddleware",
]

ROOT_URLCONF = "config.urls"

TEMPLATES = [
    {
        "BACKEND": "django.template.backends.django.DjangoTemplates",
        "DIRS": [BASE_DIR / "templates"],
        "APP_DIRS": True,
        "OPTIONS": {
            "context_processors": [
                "django.template.context_processors.debug",
                "django.template.context_processors.request",
                "django.contrib.auth.context_processors.auth",
                "django.contrib.messages.context_processors.messages",
            ],
        },
    },
]

WSGI_APPLICATION = "config.wsgi.application"


# Database
# https://docs.djangoproject.com/en/3.2/ref/settings/#databases

DATABASES = {
    "default": {
        "ENGINE": "django.db.backends.postgresql",
        "NAME": env("PORTAL_DB_DATABASE", default="portal"),
        "USER": env("PORTAL_DB_USER", default="portal"),
        "PASSWORD": env("PORTAL_DB_PASSWORD", default="portal"),
        "HOST": env("PORTAL_DB_HOST", default="localhost"),
        "PORT": env("PORTAL_DB_PORT", default="5432"),
    }
}


# User substitution
# https://docs.djangoproject.com/en/3.2/topics/auth/customizing/#auth-custom-user
AUTH_USER_MODEL = "portal.User"

# Password validation
# https://docs.djangoproject.com/en/3.2/ref/settings/#auth-password-validators

AUTH_PASSWORD_VALIDATORS = [
    {
        "NAME": "django.contrib.auth.password_validation.UserAttributeSimilarityValidator",
    },
    {
        "NAME": "django.contrib.auth.password_validation.MinimumLengthValidator",
    },
    {
        "NAME": "django.contrib.auth.password_validation.CommonPasswordValidator",
    },
    {
        "NAME": "django.contrib.auth.password_validation.NumericPasswordValidator",
    },
]


# Internationalization
# https://docs.djangoproject.com/en/3.2/topics/i18n/

LANGUAGE_CODE = "en-us"

TIME_ZONE = "America/Edmonton"

USE_I18N = True

USE_L10N = True

USE_TZ = True


# Static files (CSS, JavaScript, Images)
# https://docs.djangoproject.com/en/3.2/howto/static-files/

STATIC_URL = "/djangostatic/"
STATIC_ROOT = os.path.join(BASE_DIR, "djangostatic/")

MEDIA_URL = "/media/"
MEDIA_ROOT = os.path.join(BASE_DIR, "build", "media")

# Default primary key field type
# https://docs.djangoproject.com/en/3.2/ref/settings/#default-auto-field

DEFAULT_AUTO_FIELD = "django.db.models.BigAutoField"

if DEBUG:
    CORS_ALLOW_ALL_ORIGINS = True
else:
    frontend_origin = (
        f"{parsed_frontend_base_url.scheme}://{parsed_frontend_base_url.netloc}"
    )
    CORS_ALLOWED_ORIGINS = [frontend_origin]

REST_FRAMEWORK = {
    "DEFAULT_SCHEMA_CLASS": "drf_spectacular.openapi.AutoSchema",
    "DEFAULT_AUTHENTICATION_CLASSES": [
        "rest_framework.authentication.TokenAuthentication",
    ],
    "DEFAULT_RENDERER_CLASSES": [
        "rest_framework.renderers.JSONRenderer",
    ]
    + (["rest_framework.renderers.BrowsableAPIRenderer"] if DEBUG else []),
}

SPECTACULAR_SETTINGS = {
    "TITLE": "Capstone Projects Portal API",
    "DESCRIPTION": "This is the API for the Capstone Projects Portal.",
    "VERSION": "1.0.0",
}

# Setting for email functionality
EMAIL_BACKEND = "django.core.mail.backends.smtp.EmailBackend"
EMAIL_HOST = env("EMAIL_HOST", default="")
EMAIL_PORT = 587
EMAIL_USE_TLS = True
EMAIL_HOST_USER = env("EMAIL_HOST_USER", default="")
EMAIL_HOST_PASSWORD = env("EMAIL_HOST_PASSWORD", default="")

# GitHub OAuth2 keys
GITHUB_CLIENT_ID = env("GITHUB_CLIENT_ID", default="")
GITHUB_CLIENT_SECRET = env("GITHUB_CLIENT_SECRET", default="")

# Test runner
TEST_RUNNER = "config.test_runner.TestRunner"
