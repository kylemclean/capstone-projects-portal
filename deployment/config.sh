#!/usr/bin/env bash

# Path to root directory of project
PROJECT_DIR=/home/ubuntu/cmput401-portal

# Path to the nginx config for the portal site
PORTAL_SITE_CONFIG=/etc/nginx/sites-available/portal

# Path to the gunicorn.service file
GUNICORN_SERVICE=/etc/systemd/system/gunicorn.service
