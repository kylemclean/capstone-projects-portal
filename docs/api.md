# API Documentation

To read the API documentation, start the backend and go to `/api/docs/` in a web browser.
