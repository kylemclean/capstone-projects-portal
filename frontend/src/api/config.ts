import { AxiosRequestConfig } from "axios"

// development config
const devConfig: AxiosRequestConfig = {
    baseURL: process.env.REACT_APP_API_BASE_URL ?? "http://localhost:8000/api",
    timeout: 3000,
}

// production config
const prodConfig: AxiosRequestConfig = {
    baseURL: process.env.REACT_APP_API_BASE_URL ?? "/api",
    timeout: 3000,
}

export default process.env.NODE_ENV === "production" ? prodConfig : devConfig

export const githubAuthorizeUrl =
    process.env.REACT_APP_GITHUB_AUTHORIZE_URL ??
    "https://github.com/login/oauth/authorize"
