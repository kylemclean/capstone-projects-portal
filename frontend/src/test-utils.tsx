import * as React from "react"
import { useLocation, useParams } from "react-router-dom"
import { screen, waitFor } from "@testing-library/react"
import { useCurrentUser } from "./hooks"

/**
 * Component to display the login status of the user
 */
export const LoginStatus = (): JSX.Element => {
    const currentUser = useCurrentUser()
    return (
        <div>
            {currentUser.logged_in
                ? `Logged in as ${currentUser.name}`
                : "Not logged in"}
        </div>
    )
}

/**
 * Mock home page that shows the login status
 */
export const MockHomePage = (): JSX.Element => (
    <div>
        Home page
        <LoginStatus />
    </div>
)

/**
 * Mock profile page that shows user ID
 */
export const MockProfilePage = (): JSX.Element => {
    const { id } = useParams<{ id: string }>()
    return <div>Profile page for {id}</div>
}

/**
 * Component that displays each query parameter in a list of "key=value" pairs
 * Requires the use of the `useLocation` hook
 */
export const QueryParamsDisplay = (): JSX.Element => {
    const { search } = useLocation()
    const params = new URLSearchParams(search)
    return (
        <div>
            <ul>
                {Array.from(params.entries()).map(([key, value]) => (
                    <li key={key}>
                        {key}={value}
                    </li>
                ))}
            </ul>
        </div>
    )
}

/**
 * Waits for a query parameter `name` to be set to `value`, or unset.
 * Requires a `QueryParamsDisplay` component to be rendered on the screen.
 * @param name the name of the query parameter
 * @param value the value of the query parameter, or `null` if it should not be present
 */
export const waitForQueryParam = async (
    name: string,
    value: string | null
): Promise<void> =>
    waitFor(() => {
        const text = `${name}=${value}`
        if (value !== null) screen.getByText(text)
        else expect(screen.queryByText(text)).toBeNull()
    })

/**
 * Mock login page that shows where it will redirect to after login
 */
export const MockLoginPage = (): JSX.Element => {
    const location = useLocation()
    const next = new URLSearchParams(location.search).get("next") ?? "/"
    return <div>Login page, will redirect to {next}</div>
}
