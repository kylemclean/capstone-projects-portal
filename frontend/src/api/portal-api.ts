import axios, {
    AxiosError,
    AxiosInstance,
    AxiosRequestConfig,
    AxiosResponse,
} from "axios"
import { Dispatch } from "react"
import ClientOrg from "../models/client-org"
import User from "../models/user"
import { axiosConfig } from "./config"
import Project from "../models/project"
import ProposalForm from "../models/proposal-form"
import {
    ActivateRequest,
    ActivateResult,
    LoginResult,
    LoginWithEmailAndPasswordRequest,
    LoginWithOAuth2ProviderRequest,
    RequestPasswordResetRequest,
    ResetPasswordRequest,
    ResetPasswordResult,
} from "../models/login"
import CurrentUserInfo from "../models/current-user-info"
import Action from "../global-state/action"
import ImportCsvResponse from "../models/import"

/**
 * Returns an appropriate error message for a given error object.
 * @param error The error object.
 * @returns The appropriate error message, or "Unknown error" if the error is of an unknown type or has no message.
 */
const errorMessageOf = (
    error: AxiosError<{ error: string }> | Error | unknown
): string =>
    (error as AxiosError<{ error: string }>)?.response?.data?.error ||
    (error as Error)?.message ||
    "Unknown error"

const onLoginResult = (result: LoginResult, dispatch: Dispatch<Action>) => {
    if (result.success === true) {
        localStorage.setItem("token", result.token)
        dispatch({ type: "SET_CURRENT_USER", value: result.user })
    }
}

const handleLoginResponsePromise = (
    promise: Promise<AxiosResponse<LoginResult>>,
    dispatch: Dispatch<Action>
): Promise<LoginResult> =>
    promise
        .then((response) => {
            onLoginResult(response.data, dispatch)
            return response.data
        })
        .catch((error) => {
            const result: LoginResult = {
                success: false,
                error: errorMessageOf(error),
            }
            onLoginResult(result, dispatch)
            return result
        })

export default class PortalApi {
    axiosInstance: AxiosInstance

    constructor(axiosRequestConfig: AxiosRequestConfig, logResponses: boolean) {
        // reusable AxiosInstance
        this.axiosInstance = axios.create(axiosRequestConfig)

        // Interceptor to include token in all requests if it is set
        this.axiosInstance.interceptors.request.use((requestConfig) => {
            const token = localStorage.getItem("token")
            if (token !== null)
                requestConfig.headers.Authorization = `Token ${token}`

            return requestConfig
        })

        // log all errors
        if (logResponses) {
            this.axiosInstance.interceptors.response.use(
                (response) =>
                    // Any status code that lie within the range of 2xx cause this function to trigger
                    // console.log(response)
                    response,
                (error) => {
                    // Any status codes that falls outside the range of 2xx cause this function to trigger
                    console.error(error)

                    // Invalidate token if it results in HTTP 401
                    if (
                        error.request?.status === 401 &&
                        localStorage.getItem("token") !== null
                    ) {
                        localStorage.removeItem("token")
                        // Reload page to remove stale information
                        window.location.reload()
                        return
                    }

                    throw error
                }
            )
        }
    }

    getClientOrgs = async (): Promise<ClientOrg[]> =>
        this.axiosInstance
            .get<ClientOrg[]>("/orgs/")
            .then((response) => response.data)

    getClientOrg = async (id: string): Promise<ClientOrg> =>
        this.axiosInstance
            .get<ClientOrg>(`/orgs/${id}/`)
            .then((response) => response.data)

    editClient = async (clientId: string, data: FormData): Promise<void> =>
        this.axiosInstance
            .patch<FormData>(`/orgs/${clientId}/`, data)
            .then(() => {})

    getUser = async (id: string): Promise<User> =>
        this.axiosInstance
            .get<User>(`/users/${id}/`)
            .then((response) => response.data)

    getImage = async (image_url: string | BlobPart): Promise<Response> =>
        this.axiosInstance
            .get<Response>(`${image_url}`)
            .then((response) => response)
            .catch((error) => error.response)

    getCurrentUserInfo = async (): Promise<CurrentUserInfo> =>
        this.axiosInstance
            .get<CurrentUserInfo>("/users/me/")
            .then((response) => response.data)

    getProjects = async (homePageOnly = false): Promise<Project[]> =>
        this.axiosInstance
            .get<Project[]>(`/projects/?home_page=${homePageOnly}`)
            .then((response) => response.data)

    getProject = async (id: string): Promise<Project> =>
        this.axiosInstance
            .get<Project>(`/projects/${id}/`)
            .then((response) => response.data)

    createProposal = async (inputData: ProposalForm): Promise<void> =>
        this.axiosInstance
            .post<ProposalForm>("/proposals/", inputData)
            .then(() => {})

    editProfile = async (userId: string, inputData: FormData): Promise<void> =>
        this.axiosInstance
            .patch<FormData>(`/users/${userId}/`, inputData)
            .then(() => {})

    editProject = async (
        projectId: string,
        projectData: FormData
    ): Promise<void> =>
        this.axiosInstance
            .patch<FormData>(`/projects/${projectId}/`, projectData)
            .then(() => {})

    validateCsv = async (csvFile: File): Promise<ImportCsvResponse> => {
        const formData = new FormData()
        formData.append("file", csvFile)
        return this.axiosInstance
            .post<FormData, AxiosResponse<ImportCsvResponse>>(
                "/csv/validate/",
                formData
            )
            .then((response) => response.data)
    }

    importCsv = async (csvFile: File): Promise<ImportCsvResponse> => {
        const formData = new FormData()
        formData.append("file", csvFile)
        return this.axiosInstance
            .post<FormData, AxiosResponse<ImportCsvResponse>>(
                "/csv/import/",
                formData
            )
            .then((response) => response.data)
    }

    loginWithEmailAndPassword = async (
        inputData: LoginWithEmailAndPasswordRequest,
        dispatch: Dispatch<Action>
    ): Promise<LoginResult> =>
        handleLoginResponsePromise(
            this.axiosInstance.post<
                LoginWithEmailAndPasswordRequest,
                AxiosResponse<LoginResult>
            >("/login/email/", inputData),
            dispatch
        )

    loginWithOAuth2Provider = async (
        inputData: LoginWithOAuth2ProviderRequest,
        dispatch: Dispatch<Action>
    ): Promise<LoginResult> =>
        handleLoginResponsePromise(
            this.axiosInstance.post<
                LoginWithOAuth2ProviderRequest,
                AxiosResponse<LoginResult>
            >("/login/oauth2/", inputData),
            dispatch
        )

    logout = async (dispatch: Dispatch<Action>): Promise<void> => {
        localStorage.removeItem("token")
        dispatch({ type: "SET_CURRENT_USER", value: { logged_in: false } })
    }

    activateAccount = async (
        inputData: ActivateRequest,
        dispatch: Dispatch<Action>
    ): Promise<ActivateResult> =>
        handleLoginResponsePromise(
            this.axiosInstance.post<
                ActivateRequest,
                AxiosResponse<ActivateResult>
            >("/activate/", inputData),
            dispatch
        )

    resetPassword = async (
        inputData: ResetPasswordRequest,
        dispatch: Dispatch<Action>
    ): Promise<ResetPasswordResult> =>
        handleLoginResponsePromise(
            this.axiosInstance.post<
                ResetPasswordRequest,
                AxiosResponse<ResetPasswordResult>
            >("/reset-password/", inputData),
            dispatch
        )

    requestPasswordReset = async (
        inputData: RequestPasswordResetRequest
    ): Promise<void> =>
        this.axiosInstance
            .post<RequestPasswordResetRequest>(
                "/request-password-reset/",
                inputData
            )
            .then(() => {})

    invalidateOtherSessions = async (
        dispatch: Dispatch<Action>
    ): Promise<LoginResult> =>
        handleLoginResponsePromise(
            this.axiosInstance.post<void, AxiosResponse<LoginResult>>(
                "/logout-all/"
            ),
            dispatch
        )

    logAllData = async (): Promise<void> => {
        console.log("%c Client Orgs:", "font-size: 16px")
        console.table(await this.getClientOrgs())
        console.log("%c Projects:", "font-size: 16px")
        console.table(await this.getProjects())
    }
}

const logResponses = false
export const portalApiInstance = new PortalApi(axiosConfig, logResponses)
