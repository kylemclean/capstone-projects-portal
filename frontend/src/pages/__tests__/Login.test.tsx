import * as React from "react"

// import API mocking utilities from Mock Service Worker
import { http } from "msw"
import { setupServer } from "msw/node"

// import react-testing methods
import { render, fireEvent, waitFor, screen } from "@testing-library/react"

// the component to test
import { MemoryRouter, Route } from "react-router-dom"
import LoginPage from "../Login"

import {
    LoginResult,
    LoginWithEmailAndPasswordRequest,
} from "../../models/login"
import GlobalStateProvider from "../../global-state/provider"
import State from "../../global-state/state"
import { LoginStatus, MockHomePage } from "../../test-utils"
import { axiosConfig } from "../../api/config"

// Test user data
const userEmail = "user@example.com"
const userPassword = "correctpassword"

// Response returned from /api/login/email/ when the user fails to log in
const unsuccessfulEmailLoginResponse: LoginResult = {
    success: false,
    error: "Incorrect email or password",
}

// Response returned from /api/login/email/ when the user logs in successfully
const successfulEmailLoginResponse: LoginResult = {
    success: true,
    token: "THISISTHETOKEN",
    user: {
        logged_in: true,
        is_superuser: false,
        has_password: true,
        id: "1fc66da3-5f63-4dd8-aaac-425a18804cdc",
        name: "Guy User",
        image: "https://example.com/image.jpg",
    },
}

// Mock request password reset page
const MockRequestPasswordResetPage = () => (
    <div>Request password reset page</div>
)

// Mock some other page for testing next URL redirection, shows login status
const MockSomeOtherPage = () => (
    <div>
        Some other page
        <LoginStatus />
    </div>
)

const server = setupServer(
    // Mock /api/login/email response
    http.post(`${axiosConfig.baseURL}/login/email/`, async ({ request }) => {
        const { email, password } =
            (await request.json()) as LoginWithEmailAndPasswordRequest

        if (email === userEmail && password === userPassword)
            return Response.json(successfulEmailLoginResponse)

        return Response.json(unsuccessfulEmailLoginResponse)
    })
)

beforeAll(() => server.listen())
afterEach(() => server.resetHandlers())

/**
 * Render the login page component in a router with a GlobalStateProvider.
 * @param initialState the initial State of the global state. If undefined, it will be the default state.
 * @param nextUrl the next url to navigate to after logging in. If undefined, it will be the default url.
 * @returns return value of the render function
 */
const renderLoginPage = (initialState?: State, nextUrl?: string) => {
    // Render the login page component along with the MockHomePage
    // in a router with a GlobalStateProvider
    const url = `/login${nextUrl ? `?next=${nextUrl}` : ""}`
    const renderReturnValue = render(
        <GlobalStateProvider initialState={initialState}>
            <MemoryRouter initialEntries={[url]}>
                <Route path="/login" exact component={LoginPage} />
                <Route
                    path="/request-password-reset"
                    exact
                    component={MockRequestPasswordResetPage}
                />
                <Route
                    path="/some-other-page"
                    exact
                    component={MockSomeOtherPage}
                />
                <Route path="/" exact component={MockHomePage} />
            </MemoryRouter>
        </GlobalStateProvider>
    )

    return renderReturnValue
}

/**
 * @returns the email, password, and login button elements
 */
const getLoginFormControls = () => {
    const emailInput = screen.getByLabelText("Email")
    const passwordInput = screen.getByLabelText("Password")
    const loginButton = screen.getByRole("button", { name: "Log in" })

    return { emailInput, passwordInput, loginButton }
}

it("matches snapshot", async () => {
    const { container } = renderLoginPage()
    expect(container).toMatchSnapshot()
})

it("focuses email field on load", async () => {
    renderLoginPage()
    const { emailInput } = getLoginFormControls()

    // Check that the email input is focused
    expect(emailInput).toHaveFocus()
})

it("redirects to home page if already logged in and next URL is unspecified", async () => {
    renderLoginPage({
        currentUser: {
            logged_in: true,
            is_superuser: false,
            has_password: true,
            id: "1fc66da3-5f63-4dd8-aaac-425a18804cdc",
            name: "Guy User",
            image: "https://example.com/image.jpg",
        },
    })

    // Wait for the redirect to the home page
    await waitFor(() => screen.getByText("Home page"))

    // Check that the user is logged in
    expect(screen.getByText("Logged in as Guy User")).toBeInTheDocument()
})

it("redirects to next URL if already logged in and it is specified", async () => {
    renderLoginPage(
        {
            currentUser: {
                logged_in: true,
                is_superuser: false,
                has_password: true,
                id: "1fc66da3-5f63-4dd8-aaac-425a18804cdc",
                name: "Guy User",
                image: "https://example.com/image.jpg",
            },
        },
        "/some-other-page"
    )

    // Wait for the redirect to some other page
    await waitFor(() => screen.getByText("Some other page"))

    // Check that the user is logged in
    expect(screen.getByText("Logged in as Guy User")).toBeInTheDocument()
})

it("handles successful login with correct email and password and redirects to home page when next URL is unspecified", async () => {
    renderLoginPage()
    const { emailInput, passwordInput, loginButton } = getLoginFormControls()

    // Fill in the email and password fields
    fireEvent.change(emailInput, { target: { value: userEmail } })
    fireEvent.change(passwordInput, { target: { value: userPassword } })

    // Click the login button
    fireEvent.click(loginButton)

    // Assert that the login button is disabled
    expect(loginButton).toBeDisabled()

    // Assert that the home page is shown
    await waitFor(() => screen.getByText("Home page"))

    // Assert that the user is logged in
    expect(screen.getByText("Logged in as Guy User")).toBeInTheDocument()
})

it("handles successful login with correct email and password and redirects to next URL when it is specified", async () => {
    renderLoginPage(undefined, "/some-other-page")
    const { emailInput, passwordInput, loginButton } = getLoginFormControls()

    // Fill in the email and password fields
    fireEvent.change(emailInput, { target: { value: userEmail } })
    fireEvent.change(passwordInput, { target: { value: userPassword } })

    // Click the login button
    fireEvent.click(loginButton)

    // Assert that the login button is disabled
    expect(loginButton).toBeDisabled()

    // Assert that some other page is shown
    await waitFor(() => screen.getByText("Some other page"))

    // Assert that the user is logged in
    expect(screen.getByText("Logged in as Guy User")).toBeInTheDocument()
})

it("handles failed login with incorrect email and password", async () => {
    renderLoginPage()
    const { emailInput, passwordInput, loginButton } = getLoginFormControls()

    // Fill in the email and password fields
    fireEvent.change(emailInput, { target: { value: "wrong@example.com" } })
    fireEvent.change(passwordInput, { target: { value: "wrongpassword" } })

    // Click the login button
    fireEvent.click(loginButton)

    // Assert that the login button is disabled
    expect(loginButton).toBeDisabled()

    // Assert that an error is shown
    await waitFor(() => screen.getByText("Incorrect email or password"))

    // Assert that the login button is reenabled
    expect(loginButton).toBeEnabled()
})

it("displays error when email field is empty", async () => {
    renderLoginPage()
    const { emailInput, passwordInput, loginButton } = getLoginFormControls()

    // Fill in the password field
    fireEvent.change(passwordInput, { target: { value: userPassword } })

    // Click the login button
    fireEvent.click(loginButton)

    // Assert that an error is displayed on the email field
    await waitFor(() => screen.getByText("Email is required"))

    // Assert that the field is marked with aria-invalid and aria-describedby
    expect(emailInput).toHaveAttribute("aria-invalid", "true")
    expect(emailInput).toHaveAttribute("aria-describedby", "email-helper-text")

    // Assert that the login button is still enabled
    expect(loginButton).toBeEnabled()
})

it("displays error when password field is empty", async () => {
    renderLoginPage()
    const { emailInput, passwordInput, loginButton } = getLoginFormControls()

    // Fill in the email field
    fireEvent.change(emailInput, { target: { value: userEmail } })

    // Click the login button
    fireEvent.click(loginButton)

    // Assert that an error is displayed on the password field
    await waitFor(() => screen.getByText("Password is required"))

    // Assert that the field is marked with aria-invalid and aria-describedby
    expect(passwordInput).toHaveAttribute("aria-invalid", "true")
    expect(passwordInput).toHaveAttribute(
        "aria-describedby",
        "password-helper-text"
    )

    // Assert that the login button is still enabled
    expect(loginButton).toBeEnabled()
})

it("displays error when both email and password fields are empty", async () => {
    renderLoginPage()
    const { emailInput, passwordInput, loginButton } = getLoginFormControls()

    // Click the login button
    fireEvent.click(loginButton)

    // Assert that an error is displayed on both fields
    await waitFor(() => screen.getByText("Email is required"))
    await waitFor(() => screen.getByText("Password is required"))

    // Assert that the fields are marked with aria-invalid and aria-describedby
    expect(emailInput).toHaveAttribute("aria-invalid", "true")
    expect(emailInput).toHaveAttribute("aria-describedby", "email-helper-text")
    expect(passwordInput).toHaveAttribute("aria-invalid", "true")
    expect(passwordInput).toHaveAttribute(
        "aria-describedby",
        "password-helper-text"
    )

    // Assert that the login button is still enabled
    expect(loginButton).toBeEnabled()
})

it("redirects to request password reset page when clicking reset password link", async () => {
    renderLoginPage()

    // Click the reset password link
    fireEvent.click(screen.getByRole("link", { name: "Reset password" }))

    // Assert that the request password reset page is shown
    await waitFor(() => screen.getByText("Request password reset page"))
})

afterAll(() => server.close())
