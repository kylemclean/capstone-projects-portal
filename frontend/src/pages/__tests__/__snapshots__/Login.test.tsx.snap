// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<div>
  <div
    class="MuiContainer-root MuiContainer-maxWidthSm css-cuefkz-MuiContainer-root"
  >
    <div
      class="MuiBox-root css-zitybv"
    >
      <div
        class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 css-i0bk1x-MuiPaper-root"
      >
        <h1
          class="MuiTypography-root MuiTypography-h4 MuiTypography-alignCenter MuiTypography-gutterBottom css-1trk376-MuiTypography-root"
        >
          Log in
        </h1>
        <button
          class="MuiButton-root MuiButton-outlined MuiButton-outlinedPrimary MuiButton-sizeLarge MuiButton-outlinedSizeLarge MuiButton-fullWidth MuiButtonBase-root css-1vm8rfj-MuiButtonBase-root-MuiButton-root"
          id="continue-with-github-button"
          tabindex="0"
          type="button"
        >
          <span
            class="MuiButton-startIcon MuiButton-iconSizeLarge css-htszrh-MuiButton-startIcon"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
              data-testid="GitHubIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M12 1.27a11 11 0 00-3.48 21.46c.55.09.73-.28.73-.55v-1.84c-3.03.64-3.67-1.46-3.67-1.46-.55-1.29-1.28-1.65-1.28-1.65-.92-.65.1-.65.1-.65 1.1 0 1.73 1.1 1.73 1.1.92 1.65 2.57 1.2 3.21.92a2 2 0 01.64-1.47c-2.47-.27-5.04-1.19-5.04-5.5 0-1.1.46-2.1 1.2-2.84a3.76 3.76 0 010-2.93s.91-.28 3.11 1.1c1.8-.49 3.7-.49 5.5 0 2.1-1.38 3.02-1.1 3.02-1.1a3.76 3.76 0 010 2.93c.83.74 1.2 1.74 1.2 2.94 0 4.21-2.57 5.13-5.04 5.4.45.37.82.92.82 2.02v3.03c0 .27.1.64.73.55A11 11 0 0012 1.27"
              />
            </svg>
          </span>
          Continue with GitHub
          <span
            class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
          />
        </button>
        <div
          class="MuiDivider-root MuiDivider-fullWidth MuiDivider-withChildren css-11vmsk3-MuiDivider-root"
          role="separator"
        >
          <span
            class="MuiDivider-wrapper css-e34q8v-MuiDivider-wrapper"
          >
            or
          </span>
        </div>
        <form
          method="post"
        >
          <div
            class="MuiFormControl-root MuiFormControl-marginDense MuiFormControl-fullWidth MuiTextField-root css-1ldlpx-MuiFormControl-root-MuiTextField-root"
          >
            <label
              class="MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-outlined MuiFormLabel-root MuiFormLabel-colorPrimary Mui-focused css-1kty9di-MuiFormLabel-root-MuiInputLabel-root"
              data-shrink="true"
              for="email"
              id="email-label"
            >
              Email
            </label>
            <div
              class="MuiOutlinedInput-root MuiInputBase-root MuiInputBase-colorPrimary MuiInputBase-fullWidth Mui-focused MuiInputBase-formControl css-md26zr-MuiInputBase-root-MuiOutlinedInput-root"
            >
              <input
                aria-invalid="false"
                class="MuiOutlinedInput-input MuiInputBase-input css-1t8l2tu-MuiInputBase-input-MuiOutlinedInput-input"
                id="email"
                name="email"
                type="email"
                value=""
              />
              <fieldset
                aria-hidden="true"
                class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
              >
                <legend
                  class="css-186xcr5"
                >
                  <span>
                    Email
                  </span>
                </legend>
              </fieldset>
            </div>
            
          </div>
          <div
            class="MuiFormControl-root MuiFormControl-marginDense MuiFormControl-fullWidth MuiTextField-root css-1ldlpx-MuiFormControl-root-MuiTextField-root"
          >
            <label
              class="MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined MuiFormLabel-root MuiFormLabel-colorPrimary css-14s5rfu-MuiFormLabel-root-MuiInputLabel-root"
              data-shrink="false"
              for="password"
              id="password-label"
            >
              Password
            </label>
            <div
              class="MuiOutlinedInput-root MuiInputBase-root MuiInputBase-colorPrimary MuiInputBase-fullWidth MuiInputBase-formControl css-md26zr-MuiInputBase-root-MuiOutlinedInput-root"
            >
              <input
                aria-invalid="false"
                class="MuiOutlinedInput-input MuiInputBase-input css-1t8l2tu-MuiInputBase-input-MuiOutlinedInput-input"
                id="password"
                name="password"
                type="password"
                value=""
              />
              <fieldset
                aria-hidden="true"
                class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
              >
                <legend
                  class="css-dhh0px"
                >
                  <span>
                    Password
                  </span>
                </legend>
              </fieldset>
            </div>
            
          </div>
          <button
            class="MuiLoadingButton-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeLarge MuiButton-containedSizeLarge MuiButton-fullWidth MuiButtonBase-root css-1nelln8-MuiButtonBase-root-MuiButton-root-MuiLoadingButton-root"
            tabindex="0"
            type="submit"
          >
            Log in
            <span
              class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
            />
          </button>
        </form>
        <a
          class="MuiTypography-root MuiTypography-inherit MuiLink-root MuiLink-underlineAlways css-3k7i95-MuiTypography-root-MuiLink-root"
          href="/request-password-reset"
        >
          Reset password
        </a>
      </div>
    </div>
  </div>
</div>
`;
